File: test/language/parameterized_types/java/parameterized_types/fixed_and_variable_param/FixedAndVariableParamTest.java
package parameterized_types.fixed_and_variable_param;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.math.BigInteger;

import zserio.runtime.ZserioError;
import zserio.runtime.io.BitBuffer;
import zserio.runtime.io.BitStreamReader;
import zserio.runtime.io.BitStreamWriter;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;
import zserio.runtime.io.SerializeUtil;

public class FixedAndVariableParamTest
{
    @Test
    public void writeReadFile() throws IOException
    {
        final FixedAndVariableParam fixedAndVariableParam = createFixedAndVariableParam(ARRAY_SIZE, EXTRA_LIMIT,
                LIMIT, COLOR, ACCESS, FLOAT_VALUE);
        SerializeUtil.serializeToFile(fixedAndVariableParam, BLOB_NAME);

        final FixedAndVariableParam readFixedAndVariableParam = SerializeUtil.deserializeFromFile(
                FixedAndVariableParam.class, BLOB_NAME);
        assertEquals(fixedAndVariableParam, readFixedAndVariableParam);
    }

    @Test
    public void writeRead() throws IOException
    {
        final FixedAndVariableParam fixedAndVariableParam = createFixedAndVariableParam(ARRAY_SIZE, EXTRA_LIMIT,
                LIMIT, COLOR, ACCESS, FLOAT_VALUE);
        final BitBuffer bitBuffer = SerializeUtil.serialize(fixedAndVariableParam);
        checkFixedAndVariableParamInBitBuffer(bitBuffer, fixedAndVariableParam,
                ARRAY_SIZE, EXTRA_LIMIT, LIMIT, COLOR, ACCESS, FLOAT_VALUE);

        final FixedAndVariableParam readFixedAndVariableParam = SerializeUtil.deserialize(
                FixedAndVariableParam.class, bitBuffer);
        assertEquals(fixedAndVariableParam, readFixedAndVariableParam);
    }

    @Test
    public void writeFailureWrongArraySize() throws IOException
    {
        final FixedAndVariableParam fixedAndVariableParam = createFixedAndVariableParam(WRONG_ARRAY_SIZE,
                EXTRA_LIMIT, LIMIT, COLOR, ACCESS, FLOAT_VALUE);

        try (final BitStreamWriter writer = new ByteArrayBitStreamWriter())
        {
            assertThrows(ZserioError.class, () -> fixedAndVariableParam.write(writer));
        }
    }

    @Test
    public void writeFailureWrongExtraLimit() throws IOException
    {
        final FixedAndVariableParam fixedAndVariableParam = createFixedAndVariableParam(ARRAY_SIZE,
                EXTRA_LIMIT, LIMIT, COLOR, ACCESS, FLOAT_VALUE);
        fixedAndVariableParam.setExtraLimit(WRONG_EXTRA_LIMIT);

        try (final BitStreamWriter writer = new ByteArrayBitStreamWriter())
        {
            assertThrows(ZserioError.class, () -> fixedAndVariableParam.write(writer));
        }
    }

    @Test
    public void writeFailureWrongLimitHolder() throws IOException
    {
        final FixedAndVariableParam fixedAndVariableParam = createFixedAndVariableParam(ARRAY_SIZE,
                EXTRA_LIMIT, LIMIT, COLOR, ACCESS, FLOAT_VALUE);
        final LimitHolder limitHolder = new LimitHolder(LIMIT);
        fixedAndVariableParam.setLimitHolder(limitHolder);

        try (final BitStreamWriter writer = new ByteArrayBitStreamWriter())
        {
            assertThrows(ZserioError.class, () -> fixedAndVariableParam.write(writer));
        }
    }

    @Test
    public void writeFailureWrongColor() throws IOException
    {
        final FixedAndVariableParam fixedAndVariableParam = createFixedAndVariableParam(ARRAY_SIZE,
                EXTRA_LIMIT, LIMIT, COLOR, ACCESS, FLOAT_VALUE);
        fixedAndVariableParam.setColor(WRONG_COLOR);

        try (final BitStreamWriter writer = new ByteArrayBitStreamWriter())
        {
            assertThrows(ZserioError.class, () -> fixedAndVariableParam.write(writer));
        }
    }

    @Test
    public void writeFailureWrongAccess() throws IOException
    {
        final FixedAndVariableParam fixedAndVariableParam = createFixedAndVariableParam(ARRAY_SIZE,
                EXTRA_LIMIT, LIMIT, COLOR, ACCESS, FLOAT_VALUE);
        fixedAndVariableParam.setAccess(WRONG_ACCESS);

        try (final BitStreamWriter writer = new ByteArrayBitStreamWriter())
        {
            assertThrows(ZserioError.class, () -> fixedAndVariableParam.write(writer));
        }
    }

    @Test
    public void writeFailureWrongFloatValue() throws IOException
    {
        final FixedAndVariableParam fixedAndVariableParam = createFixedAndVariableParam(ARRAY_SIZE,
                EXTRA_LIMIT, LIMIT, COLOR, ACCESS, FLOAT_VALUE);
        fixedAndVariableParam.setFloatValue(WRONG_FLOAT_VALUE);

        try (final BitStreamWriter writer = new ByteArrayBitStreamWriter())
        {
            assertThrows(ZserioError.class, () -> fixedAndVariableParam.write(writer));
        }
    }

    private ArrayHolder createArrayHolder(int size, short extraLimit, LimitHolder limitHolder, Color color,
            Access access, float floatValue)
    {
        final BigInteger[] array = new BigInteger[size];
        for (int i = 0; i < size; ++i)
            array[i] = BigInteger.valueOf(i * i);
        final boolean hasBlack = (color == Color.BLACK);
        final boolean hasRead = (access.and(Access.Values.READ)).equals(Access.Values.READ);
        final boolean hasFloatBiggerThanOne = (floatValue > 1.0);
        return new ArrayHolder(size, extraLimit, limitHolder, color, access, floatValue, array, EXTRA_VALUE,
                hasBlack, hasRead, hasFloatBiggerThanOne);
    }

    private FixedAndVariableParam createFixedAndVariableParam(int size, short extraLimit, short limit,
            Color color, Access access, float floatValue)
    {
        final LimitHolder limitHolder = new LimitHolder(limit);
        final ArrayHolder arrayHolder = createArrayHolder(size, extraLimit, limitHolder, color, access,
                floatValue);

        return new FixedAndVariableParam(extraLimit, limitHolder, color, access, floatValue, arrayHolder);
    }

    private void checkArrayHolderInStream(BitStreamReader reader, ArrayHolder arrayHolder, int size,
            short extraLimit, LimitHolder limitHolder, Color color, Access access, float floatValue)
                    throws IOException
    {
        assertEquals(arrayHolder.getSize(), size);
        assertEquals(arrayHolder.getExtraLimit(), extraLimit);
        assertEquals(arrayHolder.getLimitHolder(), limitHolder);
        assertEquals(arrayHolder.getColor(), color);
        assertEquals(arrayHolder.getAccess(), access);
        assertEquals(arrayHolder.getFloatValue(), floatValue);

        for (int i = 0; i < size; ++i)
            assertEquals(arrayHolder.getArray()[i], reader.readVarUInt());
        assertEquals(arrayHolder.getExtraValue(), reader.readBits(3));
    }

    private void checkFixedAndVariableParamInBitBuffer(BitBuffer bitBuffer,
            FixedAndVariableParam fixedAndVariableParam, int size, short extraLimit, short limit, Color color,
            Access access, float floatValue) throws IOException
    {
        try (final BitStreamReader reader = new ByteArrayBitStreamReader(bitBuffer))
        {
            assertEquals(extraLimit, reader.readBits(8));
            assertEquals(limit, reader.readBits(8));
            assertEquals(color.getValue(), reader.readBits(2));
            assertEquals(access.getValue(), reader.readBits(4));
            assertEquals(floatValue, reader.readFloat16());
            final ArrayHolder arrayHolder = fixedAndVariableParam.getArrayHolder();
            final LimitHolder limitHolder = fixedAndVariableParam.getLimitHolder();
            checkArrayHolderInStream(reader, arrayHolder, size, extraLimit, limitHolder, color, access,
                    floatValue);
        }
    }

    private static final String BLOB_NAME = "fixed_and_variable_param.blob";

    private static final int ARRAY_SIZE = 1000;
    private static final int WRONG_ARRAY_SIZE = 1001;
    private static final byte EXTRA_VALUE = 0x05;
    private static final short EXTRA_LIMIT = 0x05;
    private static final short WRONG_EXTRA_LIMIT = 0x06;
    private static final short LIMIT = 0x06;
    private static final Color COLOR = Color.BLACK;
    private static final Color WRONG_COLOR = Color.WHITE;
    private static final Access ACCESS = Access.Values.READ;
    private static final Access WRONG_ACCESS = Access.Values.WRITE;
    private static final float FLOAT_VALUE = 2.0f;
    private static final float WRONG_FLOAT_VALUE = 1.0f;
}

File: test/extensions/language/parameterized_types/cpp/FixedAndVariableParamTest.cpp
#include "gtest/gtest.h"
#include "parameterized_types/fixed_and_variable_param/FixedAndVariableParam.h"
#include "zserio/SerializeUtil.h"

namespace parameterized_types
{
namespace fixed_and_variable_param
{

using allocator_type = FixedAndVariableParam::allocator_type;
template <typename T>
using vector_type = zserio::vector<T, allocator_type>;

class FixedAndVariableParamTest : public ::testing::Test
{
protected:
    void fillArrayHolder(ArrayHolder& arrayHolder, uint32_t size, uint8_t extraLimit, LimitHolder& limitHolder,
            Color color, Access access, float floatValue)
    {
        auto& array = arrayHolder.getArray();
        array.clear();
        for (uint32_t i = 0; i < size; ++i)
        {
            array.push_back(static_cast<uint32_t>(i * i));
        }
        arrayHolder.setExtraValue(EXTRA_VALUE);
        arrayHolder.setHasBlack(color == Color::BLACK);
        arrayHolder.setHasRead((access & Access::Values::READ) == Access::Values::READ);
        arrayHolder.setHasFloatBiggerThanOne(floatValue > 1.0F);

        arrayHolder.initialize(size, extraLimit, limitHolder, color, access, floatValue);
    }

    void fillFixedAndVariableParam(FixedAndVariableParam& fixedAndVariableParam, uint32_t size,
            uint8_t extraLimit, uint8_t limit, Color color, Access access, float floatValue)
    {
        fixedAndVariableParam.setExtraLimit(extraLimit);

        LimitHolder& limitHolder = fixedAndVariableParam.getLimitHolder();
        limitHolder.setLimit(limit);

        fixedAndVariableParam.setColor(color);
        fixedAndVariableParam.setAccess(access);
        fixedAndVariableParam.setFloatValue(floatValue);

        ArrayHolder arrayHolder;
        fillArrayHolder(arrayHolder, size, extraLimit, limitHolder, color, access, floatValue);
        fixedAndVariableParam.setArrayHolder(arrayHolder);
    }

    void checkArrayHolderInBitStream(zserio::BitStreamReader& reader, const ArrayHolder& arrayHolder,
            uint32_t size, uint8_t extraLimit, const LimitHolder& limitHolder, Color color, Access access,
            float floatValue)
    {
        ASSERT_EQ(arrayHolder.getSize(), size);
        ASSERT_EQ(arrayHolder.getExtraLimit(), extraLimit);
        ASSERT_EQ(&arrayHolder.getLimitHolder(), &limitHolder);
        ASSERT_EQ(arrayHolder.getColor(), color);
        ASSERT_EQ(arrayHolder.getAccess(), access);
        ASSERT_EQ(arrayHolder.getFloatValue(), floatValue);

        for (uint32_t i = 0; i < size; ++i)
        {
            ASSERT_EQ(arrayHolder.getArray().at(i), reader.readVarUInt());
        }

        ASSERT_EQ(arrayHolder.getExtraValue(), reader.readBits(3));
        ASSERT_EQ(color == Color::BLACK, reader.readBool());
        ASSERT_EQ((access & Access::Values::READ) == Access::Values::READ, reader.readBool());
        ASSERT_EQ(floatValue > 1.0F, reader.readBool());
    }

    void checkFixedAndVariableParamInBitStream(zserio::BitStreamReader& reader,
            const FixedAndVariableParam& fixedAndVariableParam, uint32_t size, uint8_t extraLimit,
            uint8_t limit, Color color, Access access, float floatValue)
    {
        ASSERT_EQ(extraLimit, reader.readBits(8));
        ASSERT_EQ(limit, reader.readBits(8));
        ASSERT_EQ(color, zserio::valueToEnum<Color>(static_cast<uint8_t>(reader.readBits(2))));
        ASSERT_EQ(access.getValue(), reader.readBits(4));
        ASSERT_EQ(floatValue, reader.readFloat16());
        const ArrayHolder& arrayHolder = fixedAndVariableParam.getArrayHolder();
        const LimitHolder& limitHolder = fixedAndVariableParam.getLimitHolder();
        checkArrayHolderInBitStream(
                reader, arrayHolder, size, extraLimit, limitHolder, color, access, floatValue);
    }

    static const std::string BLOB_NAME;

    static const uint8_t EXTRA_VALUE;
    static const uint32_t ARRAY_SIZE;
    static const uint32_t WRONG_ARRAY_SIZE;
    static const uint8_t EXTRA_LIMIT;
    static const uint8_t WRONG_EXTRA_LIMIT;
    static const uint8_t LIMIT;
    static const Color COLOR;
    static const Color WRONG_COLOR;
    static const Access ACCESS;
    static const Access WRONG_ACCESS;
    static const float FLOAT_VALUE;
    static const float WRONG_FLOAT_VALUE;

    zserio::BitBuffer bitBuffer = zserio::BitBuffer(3 * 1024 * 8);
};

const std::string FixedAndVariableParamTest::BLOB_NAME =
        "language/parameterized_types/fixed_and_variable_param.blob";

const uint8_t FixedAndVariableParamTest::EXTRA_VALUE = 0x05;
const uint32_t FixedAndVariableParamTest::ARRAY_SIZE = 1000;
const uint32_t FixedAndVariableParamTest::WRONG_ARRAY_SIZE = 1001;
const uint8_t FixedAndVariableParamTest::EXTRA_LIMIT = 0x05;
const uint8_t FixedAndVariableParamTest::WRONG_EXTRA_LIMIT = 0x06;
const uint8_t FixedAndVariableParamTest::LIMIT = 0x06;
const Color FixedAndVariableParamTest::COLOR = Color::BLACK;
const Color FixedAndVariableParamTest::WRONG_COLOR = Color::WHITE;
const Access FixedAndVariableParamTest::ACCESS = Access::Values::READ;
const Access FixedAndVariableParamTest::WRONG_ACCESS = Access::Values::WRITE;
const float FixedAndVariableParamTest::FLOAT_VALUE = 2.0;
const float FixedAndVariableParamTest::WRONG_FLOAT_VALUE = 1.0;

TEST_F(FixedAndVariableParamTest, writeRead)
{
    FixedAndVariableParam fixedAndVariableParam;
    fillFixedAndVariableParam(
            fixedAndVariableParam, ARRAY_SIZE, EXTRA_LIMIT, LIMIT, COLOR, ACCESS, FLOAT_VALUE);

    zserio::BitStreamWriter writer(bitBuffer);
    fixedAndVariableParam.write(writer);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    checkFixedAndVariableParamInBitStream(
            reader, fixedAndVariableParam, ARRAY_SIZE, EXTRA_LIMIT, LIMIT, COLOR, ACCESS, FLOAT_VALUE);
    reader.setBitPosition(0);

    FixedAndVariableParam readFixedAndVariableParam(reader);
    ASSERT_EQ(fixedAndVariableParam, readFixedAndVariableParam);
}

TEST_F(FixedAndVariableParamTest, writeReadFile)
{
    FixedAndVariableParam fixedAndVariableParam;
    fillFixedAndVariableParam(
            fixedAndVariableParam, ARRAY_SIZE, EXTRA_LIMIT, LIMIT, COLOR, ACCESS, FLOAT_VALUE);

    zserio::serializeToFile(fixedAndVariableParam, BLOB_NAME);

    const auto readFixedAndVariableParam = zserio::deserializeFromFile<FixedAndVariableParam>(BLOB_NAME);
    ASSERT_EQ(fixedAndVariableParam, readFixedAndVariableParam);
}

TEST_F(FixedAndVariableParamTest, writeFailureWrongArraySize)
{
    FixedAndVariableParam fixedAndVariableParam;
    fillFixedAndVariableParam(
            fixedAndVariableParam, WRONG_ARRAY_SIZE, EXTRA_LIMIT, LIMIT, COLOR, ACCESS, FLOAT_VALUE);

    zserio::BitStreamWriter writer1(bitBuffer);
    ASSERT_THROW(fixedAndVariableParam.write(writer1), zserio::CppRuntimeException);

    fixedAndVariableParam.initializeChildren();
    // this fails because of array size check even if parameters are correct
    zserio::BitStreamWriter writer2(bitBuffer);
    ASSERT_THROW(fixedAndVariableParam.write(writer2), zserio::CppRuntimeException);
}

TEST_F(FixedAndVariableParamTest, writeFailureWrongExtraLimit)
{
    FixedAndVariableParam fixedAndVariableParam;
    fillFixedAndVariableParam(
            fixedAndVariableParam, ARRAY_SIZE, EXTRA_LIMIT, LIMIT, COLOR, ACCESS, FLOAT_VALUE);
    fixedAndVariableParam.setExtraLimit(WRONG_EXTRA_LIMIT);

    zserio::BitStreamWriter writer(bitBuffer);
    ASSERT_THROW(fixedAndVariableParam.write(writer), zserio::CppRuntimeException);
}

TEST_F(FixedAndVariableParamTest, writeFailureWrongLimitHolder)
{
    FixedAndVariableParam fixedAndVariableParam;
    fillFixedAndVariableParam(
            fixedAndVariableParam, ARRAY_SIZE, EXTRA_LIMIT, LIMIT, COLOR, ACCESS, FLOAT_VALUE);
    LimitHolder limitHolder(LIMIT);
    ArrayHolder& arrayHolder = fixedAndVariableParam.getArrayHolder();
    arrayHolder.initialize(arrayHolder.getSize(), arrayHolder.getExtraLimit(), limitHolder,
            arrayHolder.getColor(), arrayHolder.getAccess(), arrayHolder.getFloatValue());

    zserio::BitStreamWriter writer(bitBuffer);
    ASSERT_THROW(fixedAndVariableParam.write(writer), zserio::CppRuntimeException);
}

TEST_F(FixedAndVariableParamTest, writeFailureWrongColor)
{
    FixedAndVariableParam fixedAndVariableParam;
    fillFixedAndVariableParam(
            fixedAndVariableParam, ARRAY_SIZE, EXTRA_LIMIT, LIMIT, COLOR, ACCESS, FLOAT_VALUE);
    fixedAndVariableParam.setColor(WRONG_COLOR);

    zserio::BitStreamWriter writer(bitBuffer);
    ASSERT_THROW(fixedAndVariableParam.write(writer), zserio::CppRuntimeException);
}

TEST_F(FixedAndVariableParamTest, writeFailureWrongAccess)
{
    FixedAndVariableParam fixedAndVariableParam;
    fillFixedAndVariableParam(
            fixedAndVariableParam, ARRAY_SIZE, EXTRA_LIMIT, LIMIT, COLOR, ACCESS, FLOAT_VALUE);
    fixedAndVariableParam.setAccess(WRONG_ACCESS);

    zserio::BitStreamWriter writer(bitBuffer);
    ASSERT_THROW(fixedAndVariableParam.write(writer), zserio::CppRuntimeException);
}

TEST_F(FixedAndVariableParamTest, writeFailureWrongFloatValue)
{
    FixedAndVariableParam fixedAndVariableParam;
    fillFixedAndVariableParam(
            fixedAndVariableParam, ARRAY_SIZE, EXTRA_LIMIT, LIMIT, COLOR, ACCESS, FLOAT_VALUE);
    fixedAndVariableParam.setFloatValue(WRONG_FLOAT_VALUE);

    zserio::BitStreamWriter writer(bitBuffer);
    ASSERT_THROW(fixedAndVariableParam.write(writer), zserio::CppRuntimeException);
}

} // namespace fixed_and_variable_param
} // namespace parameterized_types
